src/mechanics/integrationtypes/IntegrationTypeTensorProduct.cpp

```diff
@@ -0,0 +1,259 @@
+#include "mechanics/elements/ElementShapeFunctions.h"
+#include "math/Legendre.h"
+#include <iostream>
+
+// constructor
+template<int TDim>
+NuTo::IntegrationTypeTensorProduct<TDim>::IntegrationTypeTensorProduct(size_t nIps,
+                             NuTo::eIntegrationMethod method)
+{
+    std::vector<double> weights;
+
+    switch (method) {
+    case NuTo::eIntegrationMethod::GAUSS:
+    {
+        ComputeWeightsAndPoints1DGauss(nIps, weights, mIPts1D);
+        break;
+    }
+    case NuTo::eIntegrationMethod::LOBATTO:
+    {
+        ComputeWeightsAndPoints1DLobatto(nIps, weights, mIPts1D);
+        break;
+    }
+    default:
+        break;
+    }
+
+    int numIpsDim = 1;
+    for(int dim = 0; dim < TDim; dim++) numIpsDim *= nIps;
+
+    // calculate the tensor product integratration point coordinates and weights
+    mWeights.resize(numIpsDim);
+    mIPts.resize(numIpsDim);
+
+    for(int i = 0; i < numIpsDim; i++)
+    {
+        mWeights[i] = 1.;
+        int power = 1;
+        for(int dim = 0; dim < TDim; dim++)
+        {
+            int index = (i/power) % nIps;
+            mWeights[i] *= weights[index];
+            mIPts[i][dim] = mIPts1D[index];
+            power *= nIps;
+        }
+    }
+}
+
+template<int TDim>
+void NuTo::IntegrationTypeTensorProduct<TDim>::ComputeWeightsAndPoints1DLobatto(size_t nIps,
+            std::vector<double>& weights, std::vector<double>& points)
+{
+    points = NuTo::Math::Polynomial::LegendreDerivRoots(nIps - 1);
+    points.insert(points.begin(), -1.);
+    points.push_back(1.);
+
+    if (nIps > 1)
+    {
+        weights.push_back(2. / (nIps * (nIps - 1)));
+        for (size_t i = 1; i < nIps - 1; i++)
+        {
+            double lp = NuTo::Math::Polynomial::Legendre(nIps - 1, points[i]);
+            weights.push_back(2. / (nIps * (nIps - 1)) / (lp * lp));
+        }
+        weights.push_back(2. / (nIps * (nIps - 1)));
+    }
+    else throw MechanicsException(__PRETTY_FUNCTION__, "Ip number out of range.");
+}
+
+template<int TDim>
+void NuTo::IntegrationTypeTensorProduct<TDim>::ComputeWeightsAndPoints1DGauss(size_t nIps,
+            std::vector<double>& weights, std::vector<double>& points)
+{
+    points = NuTo::Math::Polynomial::LegendreRoots(nIps);
+
+    if (nIps > 0)
+    {
+        for (size_t i = 0; i < nIps; i++)
+        {
+            double dl = NuTo::Math::Polynomial::Legendre(nIps, points[i], 1);
+            weights.push_back(2. / (1. - points[i] * points[i]) / (dl * dl));
+        }
+    }
+    else throw MechanicsException(__PRETTY_FUNCTION__, "Ip number out of range.");
+}
+
+template<int TDim>
+Eigen::VectorXd NuTo::IntegrationTypeTensorProduct<TDim>::GetLocalIntegrationPointCoordinates(int rIpNum) const
+{
+    if (rIpNum >= 0 && rIpNum < (int)mIPts.size())
+        return mIPts[rIpNum];
+    else
+        throw MechanicsException(
+                "[NuTo::IntegrationTypeTensorProduct::GetLocalIntegrationPointCoordinates] Ip number out of range.");
```

Comment:
Use `__PRETTY_FUNCTION__` instead of manually typing the name of the function again.